fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java

```diff
@@ -673,7 +673,13 @@ public List<HdfsFileDesc> getHdfsFileDescs(String dirPath, boolean isSplittable,
         // block locations of the files in the given path in one operation.
         // The performance is better than getting status and block location one by one.
         try {
-            RemoteIterator<LocatedFileStatus> blockIterator = fileSystem.listLocatedStatus(new Path(uri.getPath()));
+            // files in hdfs may have multiple directories, so we need to list all files in hdfs recursively here
+            RemoteIterator<LocatedFileStatus> blockIterator = null;
+            if (!Config.recursive_dir_search_enabled) {
+                blockIterator = fileSystem.listLocatedStatus(new Path(uri.getPath()));
+            } else {
+                blockIterator = fileSystem.listFiles(new Path(uri.getPath()), true);
+            }
             while (blockIterator.hasNext()) {
                 LocatedFileStatus locatedFileStatus = blockIterator.next();
                 if (!isValidDataFile(locatedFileStatus)) {
```

Comment:
isValidDataFile() will skip the dir, you could check it.

Reply:
> isValidDataFile() will skip the dir, you could check it.

After check, now hdfs directory will be scanned recursively by call getHdfsFileDescs.